src/nnet3bin/nnet3-modify-learning-rates.cc

```diff
@@ -0,0 +1,186 @@
+      Input ki(cur_nnet_rxfilename, &binary_read);
+      cur_nnet.Read(ki.Stream(), binary_read);
+    }
+
+    int32 ret = 0;
+
+    // Get info about magnitude of parameter change.
+    Nnet diff_nnet(prev_nnet);
+    AddNnet(cur_nnet, -1.0, &diff_nnet);
+    int32 num_updatable = NumUpdatableComponents(diff_nnet);
+    Vector<BaseFloat> dot_prod(num_updatable);
+    ComponentDotProducts(diff_nnet, diff_nnet, &dot_prod);
+    dot_prod.ApplyPow(0.5); // take sqrt to get l2 norm of diff
+    KALDI_LOG << "Parameter differences per layer are "
+      << PrintVectorPerUpdatableComponent(prev_nnet, dot_prod);
+
+    Vector<BaseFloat> baseline_prod(num_updatable);
+    ComponentDotProducts(prev_nnet, prev_nnet, &baseline_prod);
+    baseline_prod.ApplyPow(0.5);
+    dot_prod.DivElements(baseline_prod);
+    KALDI_LOG << "Relative parameter differences per layer are "
+      << PrintVectorPerUpdatableComponent(prev_nnet, dot_prod);
+
+    // If relative parameter difference for a certain is zero, set it to the
+    // mean of the rest values.
+    int32 num_zero = 0;
+    for (int32 i = 0; i < num_updatable; i++) {
+      if (dot_prod(i) == 0.0) {
+        num_zero++;
+      }
+    }
+    
+    if (num_zero > 0) {
+      BaseFloat average_diff = dot_prod.Sum()
+        / static_cast<BaseFloat>(num_updatable - num_zero);
+      for (int32 i = 0; i < num_updatable; i++) {
+        if (dot_prod(i) == 0.0) {
+          dot_prod(i) = average_diff;
+        }
+      }
+      KALDI_LOG << "Zeros detected in the relative parameter difference "
+        << "vector, updating the vector to " << dot_prod ;
+    }
+
+    // Gets learning rates for previous neural net.
+    Vector<BaseFloat> prev_nnet_learning_rates(num_updatable),
+                      cur_nnet_learning_rates(num_updatable);
+    GetLearningRates(prev_nnet, &prev_nnet_learning_rates);
+    GetLearningRates(cur_nnet, &cur_nnet_learning_rates);
+    KALDI_LOG << "Learning rates for previous model per layer are "
+              << prev_nnet_learning_rates;
+    KALDI_LOG << "Learning rates for current model per layer are "
+              << cur_nnet_learning_rates;
+    
+    // Gets target geometric mean.
+    BaseFloat target_geometric_mean = 0.0; 
+    if (average_learning_rate == 0.0) {
+      target_geometric_mean = Exp(cur_nnet_learning_rates.SumLog()
+                                  / static_cast<BaseFloat>(num_updatable));
+    } else {
+      target_geometric_mean = average_learning_rate;
+    }
+    KALDI_ASSERT(target_geometric_mean > 0.0);
+
+    // Works out the new learning rates.  We start from the previous model;
+    // this ensures that if this program is run twice, we get consistent
+    // results even if it's overwritten the current model.
+    Vector<BaseFloat> nnet_learning_rates(prev_nnet_learning_rates);
+    nnet_learning_rates.DivElements(dot_prod);
+    KALDI_ASSERT(last_layer_factor > 0.0);
+    nnet_learning_rates(num_updatable - 1) *= last_layer_factor;
```

Comment:
It should be GetLearningRate().
And yes, still support it.
Dan


On Sun, Mar 6, 2016 at 7:47 PM, Vimal Manohar <notifications@github.com>
wrote:

> In src/nnet3bin/nnet3-modify-learning-rates.cc
> <https://github.com/kaldi-asr/kaldi/pull/560#discussion_r55152862>:
>
> > +    BaseFloat target_geometric_mean = 0.0;
> > +    if (average_learning_rate == 0.0) {
> > +      target_geometric_mean = Exp(cur_nnet_learning_rates.SumLog()
> > +                                  / static_cast<BaseFloat>(num_updatable));
> > +    } else {
> > +      target_geometric_mean = average_learning_rate;
> > +    }
> > +    KALDI_ASSERT(target_geometric_mean > 0.0);
> > +
> > +    // Works out the new learning rates.  We start from the previous model;
> > +    // this ensures that if this program is run twice, we get consistent
> > +    // results even if it's overwritten the current model.
> > +    Vector<BaseFloat> nnet_learning_rates(prev_nnet_learning_rates);
> > +    nnet_learning_rates.DivElements(dot_prod);
> > +    KALDI_ASSERT(last_layer_factor > 0.0);
> > +    nnet_learning_rates(num_updatable - 1) *= last_layer_factor;
>
> Should last_layer_factor still be supported if the same factor is inside
> the network?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/560/files#r55152862>.
>